### PR TITLE
846 lfric profile wrapper

### LINKED ISSUE #846: Add profile wrapper for LFRic timer
This wrapper might become part of LFRic code (to simplify linking issues).

The one issue is how to make it easy to compile the wrapper (till #813 or #745, which might add enough infrastructure code). For now I suggest to add two LFRic infrastructure files (timer_mod, and constants_mod) locally (which I think is all that is required for compiling a wrapper library) .

### PR DESCRIPTION
**Pull request by hiker**
Added an psydata profiling wrapper library using LFRic's timer function. I added three infrastructure files from LFRic, one slightly modified, to allow a stand-alone compilation (so the lfric timer can even be used with the gocean profile example).

One option we might need to discuss is if we want to include all lfric infrastructure modules into the library. If we do, we would have an additional, stand-alone profiling wrapper, but it might produce chaos if that library is then used with LFRic (since it would potentially have a different version of the modules). So that would only work if LFRic would include the wrapper itself.

Fixes #846.

### CODE REVIEW

**Comment by hiker** [comment]
Ready to go :)

**Comment by TeranIvy** [comment]
@hiker, you may want to take a look at the recently committed [LFRic ticket #2000](https://code.metoffice.gov.uk/trac/lfric/ticket/2000). It tweaked the timer subroutine slightly (Sam removed a filter in the timer routine which prevented routines with relatively small runtimes from being output).

**Comment by hiker** [comment]
> 
> 
> @hiker, you may want to take a look at the recently committed [LFRic ticket #2000](https://code.metoffice.gov.uk/trac/lfric/ticket/2000). It tweaked the timer subroutine slightly (Sam removed a filter in the timer routine which prevented routines with relatively small runtimes from being output).

Thanks - yes, I had to manually tweak that (but didn't commit it) to get more output for my gravity_wave examples. I also had to increase the number of subroutines that are supported. I've started the discussion in #846 about the implementation, and will update the timer to the latest version later.

**Comment by hiker** [comment]
I've updated this to the latest LFRic version of the files required form LFRic. Compiling the profile wrapper library requires 3 lfric infrastructure files (log, constants, timer_mod), though the library itself does NOT use these files, they are only required to compile the wrapper library independent of LFRic. 
It can be used in two ways:
1) You link this wrapper library with LFRic (and using the profiling transformation of course), all missing symbols are then resolved from LFRic files.
2) You can link *any* application with this wrapper library and additionally the .o files in the profile-wrapper-lfric directory (which is done with the profiling gocea example). This will then use the lfric timer (though no MPI support obviously).


We might want to check if maybe this version of timer_mod.f90 could be used in LFRic - it is mostly identical, except for a few lines like:
```
   subroutine calculate_timer_stats()
#ifndef STANDALONE_COMPILATION
     use mpi_mod, only:    get_comm_size
     use scalar_mod, only: scalar_type
#endif
     use log_mod,    only: log_event  ...
     implicit none

#ifndef STANDALONE_COMPILATION
     total_ranks = get_comm_size()
#endif
```
I just copied the version of timer_mod.F90 added here into the gravity mini apps, and it worked fine, without any changes to the build systems (because without defining STANDALONE_COMPILATION the original code is used).

File: doc/user_guide/profiling.rst
Comment (arporter) [inline doc/user_guide/profiling.rst:127]: s/an/a/

File: doc/user_guide/profiling.rst
Comment (arporter) [inline doc/user_guide/profiling.rst:126]: "...in codes other than LFRic - ..."

File: examples/gocean/eg5/Makefile
Comment (arporter) [inline examples/gocean/eg5/Makefile:99]: I'm sure we've discussed this before but is there some good reason why doing e.g. `make lfric` doesn't build the wrapper library for the lfric timer?

File: lib/profiling/Makefile
Comment (arporter) [inline lib/profiling/Makefile:49]: Nice.

File: lib/profiling/lfric/Makefile
Comment (arporter) [inline lib/profiling/lfric/Makefile:39]: Please add a comment describing the purpose of this define.

File: lib/profiling/lfric/Makefile
Comment (arporter) [inline lib/profiling/lfric/Makefile:35]: Please add a brief comment describing the purpose and functionality of this Makefile.

File: lib/profiling/lfric/README.md
Comment (arporter) [inline lib/profiling/lfric/README.md:6]: You need to say that this isn't just a wrapper library since it actually incorporates the LFRic timer code.

File: lib/profiling/lfric/constants_mod.F90
Comment (arporter) [inline lib/profiling/lfric/constants_mod.F90:4]: This needs to be modified, e.g. examples/lfric/eg4/mm_diagonal_kernel_mod.F90

Comment (arporter) [reply lib/profiling/lfric/constants_mod.F90:4]: Since we've modified it, it also then gains the 3-clause BSD license for the *modifications*. Again, see that file I've just pointed to.

File: lib/profiling/lfric/log_mod.F90
Comment (arporter) [inline lib/profiling/lfric/log_mod.F90:6]: Again, please update this text (and if you modified the file, please add the BSD).

Comment (arporter) [reply lib/profiling/lfric/log_mod.F90:6]: Will we be able to replace this file with the one being put in the test infrastructure? I'm not very comfortable with the amount of code being added here given that most of it is presumably not used by the timing library?

File: lib/profiling/lfric/profile_psy_data_mod.F90
Comment (arporter) [inline lib/profiling/lfric/profile_psy_data_mod.F90:4]: 2020 :-)

File: lib/profiling/lfric/profile_psy_data_mod.F90
Comment (arporter) [inline lib/profiling/lfric/profile_psy_data_mod.F90:35]: Please update.

File: lib/profiling/lfric/profile_psy_data_mod.F90
Comment (arporter) [inline lib/profiling/lfric/profile_psy_data_mod.F90:108]: Please update :-)

File: lib/profiling/lfric/timer_mod.F90
Comment (arporter) [inline lib/profiling/lfric/timer_mod.F90:3]: Please correct and add BSD for modifications.

**Review by arporter — CHANGES_REQUESTED** [review]
Thanks for adding this Joerg. It all works fine on my laptop :-)

I have some concerns about unnecessary code/code duplication. e.g. we gain the whole logging mod but presumably don't actually need most of it. I appreciate that this gives us another timer to add to our bow but I'm not sure how useful it is. It looks as though the overheads associated with it will be quite high. I'm therefore tempted to say we should just provide the wrapper and it will need to be compiled in with LFRic in order to work. (We'll need to provide instructions for that.)

I'm open to be persuaded so we can discuss... :-)

File: lib/profiling/lfric/log_mod.F90
Comment (hiker) [reply lib/profiling/lfric/log_mod.F90:6]: I actually agree with the code duplication (note that at the time I worked on this, there was no indication of having updated infrastructure files, and I really really like that we an actually execute and test code).

My suggestion would be to put this PR on hold, till the LFRic example is merged, which should allow us to have an executable example by using the infrastructure library (it probably needs an additional file or two, but I think that's justified).

Comment (arporter) [reply lib/profiling/lfric/log_mod.F90:6]: Yes, that sounds like a sensible plan.

File: doc/user_guide/profiling.rst
Comment (hiker) [reply doc/user_guide/profiling.rst]: Done.

File: doc/user_guide/profiling.rst
Comment (hiker) [reply doc/user_guide/profiling.rst]: Done.

File: examples/gocean/eg5/Makefile
Comment (hiker) [reply examples/gocean/eg5/Makefile:99]: This is tracked in #917, am happy to look at this once this is merged.

File: lib/profiling/lfric/Makefile
Comment (hiker) [reply lib/profiling/lfric/Makefile]: That define is gone now, since I am using unmodified LFRic files.

File: lib/profiling/lfric/Makefile
Comment (hiker) [reply lib/profiling/lfric/Makefile:35]: Done.

File: lib/profiling/lfric/README.md
Comment (hiker) [reply lib/profiling/lfric/README.md]: I've updated all the docs to (hopefully) properly describe this situation.

File: lib/profiling/lfric/constants_mod.F90
Comment (hiker) [reply lib/profiling/lfric/constants_mod.F90]: All files are now used unmodified from the infrastructure directory (only additional modification was to mpi_mod, which was already modified).

File: lib/profiling/lfric/log_mod.F90
Comment (hiker) [reply lib/profiling/lfric/log_mod.F90]: As discussed in the telco, done. This directory now only contains the psydata wrapper, all other  files are taken from the infrastructure libraries (where I added one or two missing files).

File: lib/profiling/lfric/profile_psy_data_mod.F90
Comment (hiker) [reply lib/profiling/lfric/profile_psy_data_mod.F90]: Let's do the time-warp again 👍

File: lib/profiling/lfric/profile_psy_data_mod.F90
Comment (hiker) [reply lib/profiling/lfric/profile_psy_data_mod.F90:35]: Done.

File: lib/profiling/lfric/profile_psy_data_mod.F90
Comment (hiker) [reply lib/profiling/lfric/profile_psy_data_mod.F90]: Done.

File: lib/profiling/lfric/timer_mod.F90
Comment (hiker) [reply lib/profiling/lfric/timer_mod.F90]: This file is gone.

**Comment by hiker** [comment]
As discussed I have added two different libraries to the ``profile/lfric`` directory. While this was not the original intention, my first implementation (add lfric files to ``profile/lfric`` had this feature as a side-effect, so I added this as a test to gocean/example/eg5 ... then when I moved files into the infrastructure library, I realised that using these files might be bad for LFRic, if our timer is e.g. outdated. But at this stage I just decided to keep the option to use the LFRic timer for non LFRic files, which was simple enough). I am not particularly for or against this feature, if you prefer to only have the version that needs to be linked with LFRic, I am happy to change this (it's not that the LFRic timer adds significant new features). If that's the case, just let me know if you prefer to remove using the LFRic timer in ``gocean/eg5``, or just link with the LFRic infrastructure library).

Ready for next review.

File: doc/user_guide/profiling.rst
Comment (arporter) [inline doc/user_guide/profiling.rst:135]: s/it/its/

File: examples/gocean/eg5/Makefile
Comment (arporter) [inline examples/gocean/eg5/Makefile:106]: I can build and run with this. Whoop!

File: examples/gocean/eg5/README.md
Comment (arporter) [inline examples/gocean/eg5/README.md:81]: Please could you change the copyright (below) to 2020-2021.

File: lib/profiling/Makefile
Comment (arporter) [inline lib/profiling/Makefile:43]: Please could you update the copyright date to "2020-2021" on this file.

File: lib/profiling/README.md
Comment (arporter) [inline lib/profiling/README.md:179]: Please add 'lfric' to the list of possible TARGETs at the top of this file.

File: lib/profiling/lfric/Makefile
Comment (arporter) [inline lib/profiling/lfric/Makefile:4]: 2020-2021 please.

File: lib/profiling/lfric/Makefile
Comment (arporter) [inline lib/profiling/lfric/Makefile:67]: Is this target used?

File: lib/profiling/lfric/Makefile
Comment (arporter) [inline lib/profiling/lfric/Makefile:70]: s/recompilation is/recompilation if/

File: lib/profiling/lfric/README.md
Comment (arporter) [inline lib/profiling/lfric/README.md:23]: "...contained in ../../../src/psyclone/tests/test_files/dynamo0p3/infrastructure"?

File: lib/profiling/lfric/README.md
Comment (arporter) [inline lib/profiling/lfric/README.md:33]: Probably worth saying that the output is written to `timer.txt`.

File: lib/profiling/lfric/profile_psy_data_mod.F90
Comment (arporter) [inline lib/profiling/lfric/profile_psy_data_mod.F90:4]: 2020-2021

File: lib/profiling/lfric/profile_psy_data_mod.F90
Comment (arporter) [inline lib/profiling/lfric/profile_psy_data_mod.F90:71]: Please update this comment - presumably num_pre/post_vars are not used here?

File: src/psyclone/tests/test_files/dynamo0p3/infrastructure/mpi_mod.f90
Comment (arporter) [inline src/psyclone/tests/test_files/dynamo0p3/infrastructure/mpi_mod.f90:68]: s/LFRic !/LFRic/

Comment (arporter) [reply src/psyclone/tests/test_files/dynamo0p3/infrastructure/mpi_mod.f90:68]: Please update the copyright date for the modifications too.

**Review by arporter — CHANGES_REQUESTED** [review]
Looks good Joerg - I only have very minor comments. I've been able to build and run gocean/eg5 with the lfric timing library.
Testing wise, we need to check that this does now work with the LFRic model/build system. Are you able to do that Joerg or shall we look for a Met Office reviewer?

File: doc/user_guide/profiling.rst
Comment (hiker) [reply doc/user_guide/profiling.rst]: Done.

File: examples/gocean/eg5/Makefile
Comment (hiker) [reply examples/gocean/eg5/Makefile:106]: 👍

File: examples/gocean/eg5/README.md
Comment (hiker) [reply examples/gocean/eg5/README.md:81]: Done.

File: lib/profiling/Makefile
Comment (hiker) [reply lib/profiling/Makefile:43]: I believe this should be 2019-2021 ... done.

File: lib/profiling/README.md
Comment (hiker) [reply lib/profiling/README.md:179]: Indeed.

File: lib/profiling/lfric/Makefile
Comment (hiker) [reply lib/profiling/lfric/Makefile]: Done.

File: lib/profiling/lfric/Makefile
Comment (hiker) [reply lib/profiling/lfric/Makefile]: No ... removed.

File: lib/profiling/lfric/Makefile
Comment (hiker) [reply lib/profiling/lfric/Makefile]: Done.

File: lib/profiling/lfric/README.md
Comment (hiker) [reply lib/profiling/lfric/README.md]: OK.

File: lib/profiling/lfric/README.md
Comment (hiker) [reply lib/profiling/lfric/README.md:34]: Good idea.

File: lib/profiling/lfric/profile_psy_data_mod.F90
Comment (hiker) [reply lib/profiling/lfric/profile_psy_data_mod.F90]: Done.

File: lib/profiling/lfric/profile_psy_data_mod.F90
Comment (hiker) [reply lib/profiling/lfric/profile_psy_data_mod.F90:71]: They are not used. I've updated the comment.

File: src/psyclone/tests/test_files/dynamo0p3/infrastructure/mpi_mod.f90
Comment (hiker) [reply src/psyclone/tests/test_files/dynamo0p3/infrastructure/mpi_mod.f90]: Done.

**Comment by hiker** [comment]
> Testing wise, we need to check that this does now work with the LFRic model/build system. Are you able to do that Joerg or shall we look for a Met Office reviewer?

That turned out to be a bit of an issue. It works if the region names are capped at 56 (or less) characters, and crashes if 57 or more are used (LFRic limits strings to 128 characters in the timer). The crash location has no immediate connection to psydata ... it's in a transpose function (but it only crashes when I link in psydata). Valgrind points to transpose as well (before itself states "The impossible happens" ... and it crashes). It looks to be related to storing the name in the psydata object (this%name = ...). But this should trigger an automatic allocation in Fortran (since this%name is allocatable, but unallocated).

Still debugging this 😡

**Comment by hiker** [comment]
I have a suspicion that the issue is a compiler issue with the transpose/matmul implementation. If I change ``transpose_matrix_vector_kernel_mod.F90`` to be:
```
  real(kind=r_def), dimension(ndf1, ndf2) :: trans

  do k = 0, nlayers-1
    do df = 1, ndf2
      x_e(df) = x(map2(df)+k)
    end do
    ik = (cell-1)*nlayers + k + 1
    trans = transpose(matrix(:,:,ik))
    lhs_e = matmul(trans,x_e)
    !lhs_e = matmul(transpose(matrix(:,:,ik)),x_e)
```
So I use an intermediate matrix to store the result of the transpose function. Then all works fine. Now that might not be entirely conclusive, since the gravity wave apps work when I do not use psydata, but typically gravity wave would crash in that kernel, valgrind reports invalid writes.

When I run the modified code with psydata+valgrind, it runs fine (till time step 3, then I stopped). Then original ``transpose_matrix_vector_kernel_mod.F90`` code will report an error in time step 1.

**Comment by hiker** [comment]
I recompiled gravity wave, without any transformations, and valgrind still reported:
```
gravity_wave ...
20210112163755.916+1100:INFO : gravity_wave: Updating XIOS timestep
20210112163755.918+1100:INFO : Start of timestep 1
==7583== Invalid write of size 8
==7583==    at 0x507E353: ??? (in /usr/lib/x86_64-linux-gnu/libgfortran.so.5.0.0)
==7583==    by 0x4D108C: __transpose_matrix_vector_kernel_mod_MOD_transpose_matrix_vector_code (transpose_matrix_vector_kernel_mod.F90:85)
==7583==    by 0x39F29C: __gravity_wave_alg_mod_psy_MOD_invoke_1 (gravity_wave_alg_mod_psy.f90:219)
==7583==    by 0x398625: __gravity_wave_alg_mod_MOD_gravity_wave_alg_step (gravity_wave_alg_mod.f90:260)
==7583==    by 0x453802: __step_gravity_wave_mod_MOD_step_gravity_wave (step_gravity_wave_mod.F90:40)
==7583==    by 0x45231E: __gravity_wave_driver_mod_MOD_run (gravity_wave_driver_mod.f90:219)
==7583==    by 0x481F95: MAIN__ (gravity_wave.f90:42)
==7583==    by 0x481FEF: main (gravity_wave.f90:14)
==7583==  Address 0x919a4e0 is 0 bytes after a block of size 4,096 alloc'd
==7583==    at 0x483B7F3: malloc (in /usr/lib/x86_64-linux-gnu/valgrind/vgpreload_memcheck-amd64-linux.so)
==7583==    by 0x507BBDB: ??? (in /usr/lib/x86_64-linux-gnu/libgfortran.so.5.0.0)
==7583==    by 0x4D108C: __transpose_matrix_vector_kernel_mod_MOD_transpose_matrix_vector_code (transpose_matrix_vector_kernel_mod.F90:85)
==7583==    by 0x39F29C: __gravity_wave_alg_mod_psy_MOD_invoke_1 (gravity_wave_alg_mod_psy.f90:219)
==7583==    by 0x398625: __gravity_wave_alg_mod_MOD_gravity_wave_alg_step (gravity_wave_alg_mod.f90:260)
==7583==    by 0x453802: __step_gravity_wave_mod_MOD_step_gravity_wave (step_gravity_wave_mod.F90:40)
==7583==    by 0x45231E: __gravity_wave_driver_mod_MOD_run (gravity_wave_driver_mod.f90:219)
==7583==    by 0x481F95: MAIN__ (gravity_wave.f90:42)
==7583==    by 0x481FEF: main (gravity_wave.f90:14)
```
So I would suspect that this crash is actually unrelated to PSyData and the LFRic timer wrapper, but it just affected memory usage enough to cause a crash. I don't have (somewhat recent) LFRic install on any other platform, so I can't easily check with a different compiler ;(

But from my point of view it should be ready for review.

**Comment by arporter** [comment]
All requested changes have been made. I think we do need to test it against HEAD of trunk of LFRic just to check that it does work. Shall I request an MO review?

**Review by arporter — APPROVED** [review]

**Comment by arporter** [comment]
@andrewcoughtrie has kindly offered to check that this works with the latest master of LFRic. (Andy, you'll need to modify the LFRic PSyclone optimisation script to add profiling, c.f. the PSyclone training.)

**Comment by hiker** [comment]
FWIW, I've just updated LFRic, and ran latest gungho with this library. And while it ran like a dog, I did get a valid looking timer.txt file (it did also crash my windows manager, but since I have ongoing issues with that, I don't think this is related).

**Review by andrewcoughtrie — APPROVED** [review]
I've checked out the branch and run it against the LFRic test-suite everything runs through fine.

**Comment by arporter** [comment]
> I've checked out the branch and run it against the LFRic test-suite everything runs through fine.

Thanks Andrew. Just to double check, have you tried this with e.g. auto-invoke-profiling enabled? If there are no profiling calls being inserted then this branch won't have any effect.

**Comment by andrewcoughtrie** [comment]
Having forgotten to include profiling in the run when I tested (thanks for double checking @arporter) the LFRic test suite is failing due to being unable to find `profile_psy_data_mod`. I installed PSyclone using @TeranIvy's script. I'm not sure how you want the finding of this module to be handled @hiker, or whether there are some extra steps required for all this to work, either in the PSyclone install or the usage.

**Comment by arporter** [comment]
Thanks @andrewcoughtrie. As detailed here https://psyclone.readthedocs.io/en/stable/profiling.html, it's up to the application developer to ensure that the include paths are correct when building. I'm afraid I have no clue about the LFRic build system so can't help :-)

**Comment by hiker** [comment]
Here what I do for gravity wave, iirc gungho was pretty much the same:
1) Modify psyclone flags to use ``--profile-invokes`` or ``--profile-kernels`` (any one will do for this test).
2) modify .../miniapps/gravity_wave/Makefile
   - export IGNORE_DEPENDENCIES= ...    profile_psy_data_mod
   - export EXTERNAL_STATIC_LIBRARIES = ...  psy_lfric_timer
3) copy profile_psy_data_mod.mod file to .../lfric/usr/include
4) copy libpsy_lfric_timer.afile to .../lfric/usr/lib

File: lib/profiling/lfric/profile_psy_data_mod.F90
Comment (andrewcoughtrie) [inline lib/profiling/lfric/profile_psy_data_mod.F90:84]: Character* is obsolete so the Intel complier errors out when compiling this, changed it to just compiler(*) and everything worked fine.

**Review by andrewcoughtrie — CHANGES_REQUESTED** [review]
I added the `profile_psy_data_mod.F90` file to the lfric `infrastructure/source/utilities` directory (not a permanent solution for use) which allowed me to compile and run everything on spice using both the GNU and Intel compiler. With the sole exception of the character* issue I've noted inline I'm happy everything works as it should.

File: lib/profiling/lfric/profile_psy_data_mod.F90
Comment (arporter) [reply lib/profiling/lfric/profile_psy_data_mod.F90:84]: @hiker would you mind addressing this and then I can proceed to merge? Thanks :-)

Comment (hiker) [reply lib/profiling/lfric/profile_psy_data_mod.F90:84]: *grin* I show my age there :) Which compiler version are you using? I had verified this with intel 17 and intel 21 (and did just again), and both are happy with this. I just want to be able test it (since I've used character*(*) in a few places, I'd prefer to fix them all at once).

Comment (andrewcoughtrie) [reply lib/profiling/lfric/profile_psy_data_mod.F90]: I just did a rerun myself (we use 17 as well) and i can't reproduce the error now, running the test-suite again to see if it was a specific app that flagged it up. I have checked and it was made obsolete as of F95 so at least i'm not imagining things.

Comment (andrewcoughtrie) [reply lib/profiling/lfric/profile_psy_data_mod.F90]: Ok so gets flagged when compiling everything but the infrastructure unit tests, perhaps we have a flag on that you don't? our flags are set in infrastructure/build/fortran/ifort.mk if you want to have a look.

Comment (hiker) [reply lib/profiling/lfric/profile_psy_data_mod.F90]: ``-stand f08`` flags it, but it flags even more stuff - anyway, it's merged. Thanks

**Pull request merged**
